examples/platforms/efr32/uart.c

```diff
@@ -39,155 +39,145 @@
-            sReceive.mTail = (sReceive.mTail + 1) % kReceiveBufferSize;
-        }
+        sReceiveFifo.mBuffer[sReceiveFifo.mTail] = aData[0];
+        sReceiveFifo.mTail = (sReceiveFifo.mTail + 1) % kReceiveFifoSize;
     }
-}
 
-otError otPlatUartDisable(void)
-{
-    return OT_ERROR_NOT_IMPLEMENTED;
+    UARTDRV_Receive(aHandle, aData, 1, receiveDone);
 }
 
-otError otPlatUartSend(const uint8_t *aBuf, uint16_t aBufLength)
+static void transmitDone(UARTDRV_Handle_t aHandle, Ecode_t aStatus, uint8_t *aData, UARTDRV_Count_t aCount)
 {
-    otError error = OT_ERROR_NONE;
-
-    otEXPECT_ACTION(sTransmitBuffer == NULL, error = OT_ERROR_BUSY);
-
-    sTransmitBuffer = aBuf;
-    sTransmitLength = aBufLength;
-
-exit:
-    return error;
+    sTransmitLength = 0;
 }
 
-void processReceive(void)
+static void processReceive(void)
 {
     // Copy tail to prevent multiple reads
-    uint16_t tail = sReceive.mTail;
+    uint16_t tail = sReceiveFifo.mTail;
 
     // If the data wraps around, process the first part
-    if (sReceive.mHead > tail)
+    if (sReceiveFifo.mHead > tail)
     {
-        otPlatUartReceived(sReceive.mBuffer + sReceive.mHead, kReceiveBufferSize - sReceive.mHead);
+        otPlatUartReceived(sReceiveFifo.mBuffer + sReceiveFifo.mHead, kReceiveFifoSize - sReceiveFifo.mHead);
 
         // Reset the buffer mHead back to zero.
-        sReceive.mHead = 0;
+        sReceiveFifo.mHead = 0;
     }
 
     // For any data remaining, process it
-    if (sReceive.mHead != tail)
+    if (sReceiveFifo.mHead != tail)
     {
-        otPlatUartReceived(sReceive.mBuffer + sReceive.mHead, tail - sReceive.mHead);
+        otPlatUartReceived(sReceiveFifo.mBuffer + sReceiveFifo.mHead, tail - sReceiveFifo.mHead);
 
         // Set mHead to the local tail we have cached
-        sReceive.mHead = tail;
+        sReceiveFifo.mHead = tail;
     }
 }
 
-void processTransmit(void)
+static void processTransmit(void)
 {
-    otEXPECT(sTransmitBuffer != NULL);
+    if (sTransmitBuffer != NULL && sTransmitLength == 0)
+    {
+        sTransmitBuffer = NULL;
+        otPlatUartSendDone();
+    }
+}
 
-    for (; sTransmitLength > 0; sTransmitLength--)
+otError otPlatUartEnable(void)
+{
+    UARTDRV_Init_t uartInit = USART_INIT;
+
+    sReceiveFifo.mHead = 0;
+    sReceiveFifo.mTail = 0;
+
+    UARTDRV_Init(sUartHandle, &uartInit);
+
+    for (uint8_t i = 0; i < sizeof(sReceiveBuffer); i++)
     {
-        USART_Tx(USART0, *sTransmitBuffer++);
+        UARTDRV_Receive(sUartHandle, &sReceiveBuffer[i], sizeof(sReceiveBuffer[i]), receiveDone);
     }
 
-    sTransmitBuffer = NULL;
-    otPlatUartSendDone();
+    return OT_ERROR_NONE;
+}
+
+otError otPlatUartDisable(void)
+{
+    return OT_ERROR_NOT_IMPLEMENTED;
+}
+
+otError otPlatUartSend(const uint8_t *aBuf, uint16_t aBufLength)
+{
+    otError error = OT_ERROR_NONE;
+
+    otEXPECT_ACTION(sTransmitBuffer == NULL, error = OT_ERROR_BUSY);
+
+    sTransmitBuffer = aBuf;
+    sTransmitLength = aBufLength;
+
```

Comment:
Given UARTDRV_Transmit is asynchronous and aBuf can be reclaimed after otPlatUartSend returns, we should copy aBuf contents, not just its pointer.

Reply:
Ah, just saw that layer above makes sure the buffer is not touched until this is done. Disregard.